### PR TITLE
Turn Dwarvish Miners into a Level 0 unit

### PR DESCRIPTION
**Pull request by Gothyoba**
This was proposed in the forums a while back, and I personally think its a good idea. This changes the Dwarvish Miner to a level 0 unit which can advance to a Dwarvish Ulfserker, Fighter or Guardsman. It also reduces its XP from 50 to 19. The Dwarvish Fighter's stats look like a level 0 unit, and I think it would make sense as one. These are all proposals made by various people on this thread, not myself: https://forums.wesnoth.org/viewtopic.php?t=57995

### CODE REVIEW

**Comment by ProditorMagnus** [comment]
Amla and cost need changes. They are level 1 exactly because they are too cheap, so punished by upkeep and giving enemies xp.

**Comment by Gothyoba** [comment]
I just increase the cost to 11.

**Comment by Dalas121** [comment]
Agreed that miners would make more sense as a L0.

In addition to the cost change already proposed, I'd suggest lowering their resistances down to match the scout (flat 10% instead of 20% Blade/Pierce/Impact), and buffing up their damage closer to the woodsman's 4x3, or a goblin spearman's 6x3 (5x3 if weak).

**Comment by Gothyoba** [comment]
I copied over the scout resistances and increased attacks to three.

**Comment by Gothyoba** [comment]
I just realized the unit description of the Drwavish Miner does not follow Wesnoth’s typography standards of capitalizing unit names, so I fixed as part of this PR.

**Comment by CelticMinstrel** [comment]
I always like having level 0 "citizen" type units, so I welcome this change. It's a bit unfortunate that it's happening now instead of before we moved it into core, but oh well.

**Comment by Wedge009** [comment]
I'm not an expert in game-play balancing but, for me, the main thing that comes to mind with level 0 vs level 1 is lack of zone-of-control for the former. But if I recall correctly the miners aren't intended to be used in combat, so I think that's probably fine.

**Comment by knyghtmare** [comment]
Demoting a level 1 unit to level 0 will require a follow up on the scenario balances that will need to be changed as adjustments. Such as that Gathering Materials Scenario in SoF and that North Knalga scenario in SotA. Just changing the unit in the PR wont suffice.

File: data/core/units/dwarves/Miner.cfg [hunk 1]
```diff
@@ -6,15 +6,19 @@
...
     movement=4
-    level=1
+    level=0
     alignment=neutral
-    advances_to=null
-    {AMLA_DEFAULT}
-    experience=50
-    cost=5
+    advances_to=Dwarvish Fighter,Dwarvish Guardsman,Dwarvish Ulfserker
+    experience=19
```
Comment (knyghtmare) [inline data/core/units/dwarves/Miner.cfg:18]: change this to 24 
(from Hejnewar, the Mp dev whose current github account has been locked until further notice).

File: data/core/units/dwarves/Miner.cfg [hunk 2]
```diff
@@ -6,15 +6,19 @@
...
+    [/resistance]
     movement=4
-    level=1
+    level=0
     alignment=neutral
-    advances_to=null
-    {AMLA_DEFAULT}
-    experience=50
-    cost=5
+    advances_to=Dwarvish Fighter,Dwarvish Guardsman,Dwarvish Ulfserker
```
Comment (knyghtmare) [inline data/core/units/dwarves/Miner.cfg:17]: Why not Dwarvish Scout too?

**Review by knyghtmare — CHANGES_REQUESTED** [review]

File: data/core/units/dwarves/Miner.cfg (hunk 2 above)
Comment (Gothyoba) [reply data/core/units/dwarves/Miner.cfg]: I’ve replaced the Dwarvish Ulfserker with a Dwarvish Scout, which makes more sense to me personally.

File: data/core/units/dwarves/Miner.cfg (hunk 1 above)
Comment (Gothyoba) [reply data/core/units/dwarves/Miner.cfg]: Done.

**Comment by Dalas121** [comment]
> Demoting a level 1 unit to level 0 will require a follow up on the scenario balances that will need to be changed as adjustments

I did take a quick look at this earlier - Miners are only used in 3 mainline scenarios. In SoF S4 they're gold-carrying units that the player escorts around, in SotA S20 they're flavor units that flee from the player, and in EI they're combat units for an AI side.

In EI I'm confident the proposed changes won't be an issue, and in SotA their stats seem mostly irrelevant. SoF's S4 would be slightly harder (mostly due to the cost increase), but since you only recruit a small number of miners, I don't think this would be a problem either.

**Comment by Gothyoba** [comment]
I am wondering if the advancement should be changed back to Dwarvish Ulfserker instead of Dwarvish Scout? Do these changes have any influence on gameplay in the HttT remake here? I'm not quite sure if the change to Dwarvish Scout is a good change anyways.

**Comment by Dalas121** [comment]
In the HttT rework, the player unlocks Miners at the same time they get a loyal Ulf (_Ulfdain, from the original_), so at the time it made sense for them to advance into Ulfs. Scout is also quite a bit more powerful and flexible.

So I prefer Ulf, but I don't think Scout is necessarily an issue if that's what others want.

**Comment by Gothyoba** [comment]
I've added both a Dwarvish Scout and Dwarvish Ulfserker advancement. I don't think changing scenarios is necessary here.

**Comment by CelticMinstrel** [comment]
> increased cost of the unit

??? Its level went down but its cost went _up_?

> So I prefer Ulf, but I don't think Scout is necessarily an issue if that's what others want.

Only minor downside of scout here is that it's not in the main Knalgans faction, so someone wishing to use the miner with that faction (for example, in an MP scenario) would be introducing the scout into the pool of possible units. It _might_ be better for the scout to be an "official optional advancement". Although `[modify_unit_type]` can remove an advancement, isn't it better for the _default_ state to be closest to what we'd expect people most often want?

**Comment by Gothyoba** [comment]
> > increased cost of the unit
> 
> ??? Its level went down but its cost went _up_?

5 personally feels much too low, and is only used by a small handful of units right now. I copied 11 gold from the HttT remake. While this can be changed, I think this is probably quite reasonable in my opinion. Current level 0 units vary between 4 and 15 gold in cost.

I have made the Dwarvish Scout an optional advancement macro, which does seem like a good option in my opinion. I am wondering I should also add this optional advancement in /src/game_config_manager.cpp on line 419?

**Comment by sevu** [comment]
I don't think it is a good idea to make the advancement optional. The miner isn't part of the Knalgan faction either.
The open question is what we want it to advance to and to what not.

**Comment by Gothyoba** [comment]
> I don't think it is a good idea to make the advancement optional. The miner isn't part of the Knalgan faction either. The open question is what we want it to advance to and to what not.

The otpional advancement could be removed and added as a default advancement. Personally I think Guardman, Fighter, Ulfserker and Scout are good advancements. Thunderer feels too powerful me.

**Comment by CelticMinstrel** [comment]
> I am wondering I should also add this optional advancement in /src/game_config_manager.cpp on line 419?

If I recall correctly, the line you refer to is about optional advancements that predate `[modify_unit_type]` (to warn you that you need to change how you specify the advancement or it won't work), which means new optional advancements don't need to be added there.

> I don't think it is a good idea to make the advancement optional. The miner isn't part of the Knalgan faction either.
> The open question is what we want it to advance to and to what not.

I think it is better not to make the scout a default advancement, because a scenario could easily wish to add a miner to a Knalgan side, where a scout could then upset the balance. Whether or not there should be an official optional advancement is probably not for me to decide – people can use `[modify_unit_type]` directly if they want to, after all.

**Comment by Gothyoba** [comment]
I’ve removed the advancement to Scout entirely for now. Advncing the miner is so far not very important in mainline. UMC can choose to have the scout advancement with a macro or [modify_unit_type], but ordinarily I don’t think this would be reccommended. It could be made itno an official optional advancement, though I personally don’t think that is particularly necessary.

**Review by knyghtmare — APPROVED** [review]

**Review by Dalas121 — APPROVED** [review]

**Review by sevu — APPROVED** [review]

**Comment by Gothyoba** [comment]
Does someone else still need to review this right now?

**Comment by Dalas121** [comment]
One more minor comment from me. In the unit's description, can you replace: _"They have no military training, but they can still do some damage with their picks."_ with: _"Despite their lack of military training, their weighty pickaxes can deal substantial damage."_

And Hejnewar's comment, relayed from the discord: "Didnt Knyght already conveyed what I wanted to say? I dont think anything changed."

So other than that minor unit description change, I think you should be good to merge.

**Comment by Gothyoba** [comment]
The unit description has been updated.

**Comment by moonymeric** [comment]
I am totally fine with the Thunderer being an option like 'ENABLE_ANCIENT_LICH'. Maybe that's even better. The main argument for my suggestion was that the L1 dwarf units can be roughly grouped into Fighter/Thunderer/Guardsman and Ulfserker/Scout - in terms of resistances, movement points and attack types.
From that standpoint, the Thunderer is more fitting than the Ulfserker or the Scout. Of course the sudden 18-1 ranged attack makes the Miner->Thunderer sketchy, but then there's also the Peasant->Bowman or M.Citizen->M.Hunter...

As for the Woodcutter unit, I didn't mean to request its addition to core. As you said, that would be its own topic. Rather I wanted show that even if the Miner won't advance to the Ulfserker/Scout, we still have room for a precursor in the future. The Miner doesn't need to advance into as many units as possible. But it has to be balanced against other L0 units - and be plausibly designed in terms of attack damage/numbers/type and advancements. Those are my main concerns.

Also sorry if this answer feels a bit repetitive - I just wanted to make sure we don't misunderstand each other :D

**Comment by Gothyoba** [comment]
No worries! I will probably change the advancements later. I’m not sure about an optional thunderer advancement here, but I’m open to removing Ulfserkers. Slightly buffing its attack might be reasonable, but I’m not sure the HP/resistance changes are necessary.

**Comment by moonymeric** [comment]
Keeping only the Fighter and Guardsman advancements seems like the best option. Additional advancements can always be added via [modify_unit_type], if umc authors like me wish to do so. I think we can agree there.

About the hitpoints: 25hp combined with movement_type=dwarvishfoot is simply too much, compared to what you get with the Ruffian, Peasant, Woodsman, Merman Citizen, Goblin Spearman, Walking Corpse or even the Wose Sapling. The latter three don't even get normal traits, which are more noticeable on L0 units. A resilient Miner with ~29hp would be unkillable by comparable units. This should be a good case for why the hitpoints should be nerfed.

Apart from that, I'm mostly content with the unit as it exists in 1.18 mainline. Giving the Miner a 4-3 attack might make him too strong overall, when combined with terrain evasion, hitpoints and resistances. But a 5-2 melee-pierce would be exactly right, in my opinion. The resistances are probably fine, but if balance requires it, I'm okay with nerfing them from 20% to 10%. Maybe an experienced era designer like Hejnewar can help us out here.

Updated list of proposals, by order of importance:
 1. Remove Ulfserker from advancements
 2. Nerf HP from 25 to 21.
(3. Nerf attacks from 4-3 to 5-2)
(4. Keep resistances at 20%)

**Comment by Gothyoba** [comment]
I agree with most of this, though I think the HP nerf might a bit too much, and I’m also not sure why resistance should stay at 20%.

**Comment by knyghtmare** [comment]
3 approvals and seems finalized. Will merge within the next 5 hours (end of the day, for me).

**Comment by moonymeric** [comment]
My argument was that the Miner is too strong overall. All other L0 units with the same damage suffer from penalties that the Miner lacks. These are either a lack of traits or worse stats overall. And the 4mp are compensated by movement_type=dwarvishfoot.

How exactly we balance this unit is of secondary importance to me. My idea was to simply nerf HP by 4 and keep the unit mostly as it is. If others wish to buff the damage and compensate that with a nerf to resistances... I'm okay with that too, as long the end result is still balanced. Right now, I still think that the Miner's HP are too much for his movement_type, traits, resistances and damage. No other L0 unit comes even close, except the Merman Citizen who is 1. waterbound and 2. a bit overpowered too.

Is anyone with me for further nerfing the HP from 23 to 21?

**Comment by Gothyoba** [comment]
Units like the piglet are fairly comparable here. Though much weaker in other aspects, the icemoax has 35 hitpoints. I don’t think it would be an issue to nerf it even further, but I feel it’s probably fine as is.

**Comment by knyghtmare** [comment]
Sigh...all this for a level 0...
Okay, one more day, and nerf HP to 22.

**Comment by Gothyoba** [comment]
On further thought, I think 21/22 HP is quite reasonable given the attack.

**Pull request merged**